graviti/utility/attr.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 Graviti. Licensed under MIT License.
+#
+"""Attr related class."""
+
+from itertools import chain
+from typing import (
+    Any,
+    Dict,
+    Generic,
+    Iterable,
+    Iterator,
+    Mapping,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from tensorbay.utility import ReprMixin, ReprType
+
+_T = TypeVar("_T")
+_D = TypeVar("_D")
+
+
+class _AttrDict(Generic[_T], ReprMixin):
+
+    _repr_type = ReprType.MAPPING
+
+    def __init__(self) -> None:
+        self._modules: Dict[str, _AttrDict[_T]] = {}
+        self._data: Dict[str, _T] = {}
+
+    def __len__(self) -> int:
+        return sum(map(len, self._modules)) + len(self._data)
+
+    def __iter__(self) -> Iterator[str]:
+        for key, value in self._modules.items():
+            for suffix in value:
+                yield ".".join((key, suffix))
+
+        yield from self._data
+
+    def __contains__(self, key: object) -> bool:
+        if not isinstance(key, str):
+            return False
+
+        try:
+            self._getitem(key)
+        except KeyError:
+            return False
+
+        return True
+
+    def __getitem__(self, key: str) -> _T:
+        try:
+            return self._getitem(key)
+        except KeyError:
+            return self.__missing__(key)
+
+    def __setitem__(self, key: str, value: _T) -> None:
+        try:
+            prefix, suffix = key.split(".", 1)
+        except ValueError:
+            if key in self._modules:
+                raise KeyError("Key prefix conflict") from None
+
+            self._data.__setitem__(key, value)
+            return
+
+        if prefix in self._data:
+            raise KeyError("Key prefix conflict")
+
+        module = self._modules.get(prefix)
+        if not module:
+            module = _AttrDict()
+            self._modules.__setitem__(prefix, module)
+
+        module.__setitem__(suffix, value)
+
+    def __getattr__(self, key: str) -> Union["_AttrDict[_T]", _T]:
+        try:
+            return self._data.__getitem__(key)
+        except KeyError:
+            pass
+
+        try:
+            return self._modules.__getitem__(key)
+        except KeyError:
+            raise AttributeError(
+                f"'{self.__class__.__name__}' object has no attribute '{key}'"
+            ) from None
+
+    def __dir__(self) -> Iterable[str]:
+        return chain(self._data, self._modules, super().__dir__())
+
+    def __missing__(self, key: str) -> _T:
+        raise KeyError(key)
+
+    def _getitem(self, key: str) -> _T:
+        try:
+            prefix, suffix = key.split(".", 1)
+        except ValueError:
+            return self._data.__getitem__(key)
+
+        return self._modules.__getitem__(prefix).__getitem__(suffix)
+
+    def _repr_head(self) -> str:
+        return AttrDict.__name__
+
+
+class AttrDict(_AttrDict[_T], Mapping[str, _T]):
+    """A dict which allows for attr-style access of values."""
```

Comment:
`A dict that allows attr-style access of values.`?

Reply:
copyed from another python library `AttrDict`
https://github.com/bcj/AttrDict/blob/8c1883162178a124ee29144ca7abcd83cbd9d222/attrdict/mixins.py#L159

Reply:
Use `that` instead of `which`?